File: migrate/src/java/com/tacitknowledge/util/migration/jdbc/SqlScriptMigrationTask.java
/* 
 * Copyright 2007 Tacit Knowledge LLC
 * 
 * Licensed under the Tacit Knowledge Open License, Version 1.0 (the "License");
 * you may not use this file except in compliance with the License. You may
 * obtain a copy of the License at http://www.tacitknowledge.com/licenses-1.0.
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.tacitknowledge.util.migration.jdbc;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.sql.Connection;
import java.sql.SQLException;
import java.sql.Statement;
import java.util.ArrayList;
import java.util.List;
import java.util.ListIterator;
import java.util.regex.Pattern;

import org.apache.commons.lang.StringUtils;
import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;

import com.tacitknowledge.util.migration.MigrationContext;
import com.tacitknowledge.util.migration.MigrationException;
import com.tacitknowledge.util.migration.MigrationTaskSupport;
import com.tacitknowledge.util.migration.jdbc.util.SqlUtil;
import com.tacitknowledge.util.migration.jdbc.util.SybaseUtil;

/**
 * Adaptss a SQL or DDL database patch for use with the AutoPatch framework.  
 *  
 * @author  Scott Askew (scott@tacitknowledge.com)
 */
public class SqlScriptMigrationTask extends MigrationTaskSupport
{
    /**
     * Class logger
     */
    private static Log log = LogFactory.getLog(SqlScriptMigrationTask.class);
    
    /**
     * The SQL to execute
     */
    private String sql = null;
    
    /**
     * Creates a new <code>SqlScriptMigrationTask</code>.
     */
    public SqlScriptMigrationTask()
    {
        // do nothing
    }
    
    /**
     * Creates a new <code>SqlScriptMigrationTask</code>
     *
     * @param name the name of the SQL script to execute; this is just an
     *        identifier and does not have to correspond to a file name 
     * @param level the patch level of the migration task
     * @param sql the SQL to execute
     */ 
    public SqlScriptMigrationTask(String name, int level, String sql)
    {
        setName(name);
        setLevel(new Integer(level));
        this.sql = sql;
    }
    
    /**
     * Creates a new <code>SqlScriptMigrationTask</code> containing the SQL
     * contained in the given <code>InputStream</code>.
     *
     * @param  name the name of the SQL script to execute; this is just an
     *         identifier and does not have to correspond to a file name 
     * @param  level the patch level of the migration task
     * @param  is the source of the SQL to execute
     * @throws IOException if there was problem reading the input stream
     */
    public SqlScriptMigrationTask(String name, int level, InputStream is) throws IOException
    {
        setName(name);
        setLevel(new Integer(level));
        StringBuffer sqlBuffer = new StringBuffer();
        BufferedReader buf = new BufferedReader(new InputStreamReader(is));
        String line = buf.readLine();
        while (line != null)
        {
            sqlBuffer.append(line).append("\n");
            line = buf.readLine();
        }
        sql = sqlBuffer.toString();
    }

    /** {@inheritDoc} */
    public void migrate(MigrationContext ctx) throws MigrationException
    {
        JdbcMigrationContext context = (JdbcMigrationContext) ctx;
        
        Connection conn = null;
        Statement stmt = null;
        String sqlStatement = "";
        ListIterator listIterator = null;
        try
        {
            conn = context.getConnection();
            
            // cleaning the slate before we execute the patch.
            // This was inspired by a Sybase ASE server that did not allow
            // ALTER TABLE statements in multi-statement transactions.  Instead of putting
            // a if(sybase) conditional, we decided to clean the slate for everyone.
            context.commit();

            List sqlStatements = getSqlStatements(context);
            for (listIterator = sqlStatements.listIterator(); listIterator.hasNext();)
            {
                sqlStatement = (String) listIterator.next();
                log.debug(getName() + ": Attempting to execute: " + sqlStatement);

                stmt = conn.createStatement();
                
                // handle sybase special case with illegal commands in multi
                // command transactions
                if (isSybase(context) 
                      && SybaseUtil.containsIllegalMultiStatementTransactionCommand(sqlStatement))
                {
                    log.warn("Committing current transaction since patch " + getName()
                            + " contains commands that are not allowed in multi statement"
                            + " transactions.  If the patch contains errors, this patch may"
                            + " not be rolled back cleanly.");
                    context.commit();
                    stmt.execute(sqlStatement);
                    context.commit();
                }
                else // regular case
                {
                    stmt.execute(sqlStatement);
                }

                SqlUtil.close(null, stmt, null);
            }
            
            context.commit();
        }
        catch (Exception e)
        {
            String message = getName() + ": Error running SQL at statement number "
                + listIterator.previousIndex() + " \"" + sqlStatement + "\"";
            log.error(message, e);
            
            if (e instanceof SQLException)
            {
                if (((SQLException) e).getNextException() != null)
                {
                    log.error("Chained SQL Exception", ((SQLException) e).getNextException());
                }
            }
            
            context.rollback();
            throw new MigrationException(message, e);
        }
        finally
        {
            SqlUtil.close(null, stmt, null);
        }
    }
    
    /**
     * Parses the SQL/DDL to execute and returns a list of individual statements.  For database
     * types that support mulitple statements in a single <code>Statement.execute</code> call,
     * this method will return a one-element <code>List</code> containing the entire SQL
     * file.
     * 
     * @param  context the MigrationContext, to figure out db type and if it 
     *                 can handle multiple statements at once
     * @return a list of SQL and DDL statements to execute
     */
    public List getSqlStatements(JdbcMigrationContext context)
    {
        List statements = new ArrayList();
        if (context.getDatabaseType().isMultipleStatementsSupported())
        {
            statements.add(sql);
            return statements;
        }
        
        StringBuffer currentStatement = new StringBuffer();
        boolean inQuotedString = false;
        boolean inComment = false;
        char[] sqlChars = sql.toCharArray();
        for (int i = 0; i < sqlChars.length; i++)
        {
            if (sqlChars[i] == '\n')
            {
                inComment = false;
            }
            
            if (!inComment)
            {
                switch (sqlChars[i])
                {
                    case '-' :
                    case '/' :
                        if (!inQuotedString && i + 1 < sqlChars.length
                            && sqlChars[i + 1] == sqlChars[i])
                        {
                            inComment = true;
                        }
                        else
                        {
                            currentStatement.append(sqlChars[i]);
                        }
                        break;
                    case '\'' :
                        inQuotedString = !inQuotedString;
                        currentStatement.append(sqlChars[i]);
                        break;
                    case ';' :
                        if (!inQuotedString)
                        {
                            // If we're in a stored procedure, just keep rolling
                            if (isStoredProcedure(context.getDatabaseType().getDatabaseType(),
                                                  currentStatement.toString())) 
                            {
                                currentStatement.append(sqlChars[i]);
                            }
                            else 
                            {
                                statements.add(currentStatement.toString().trim());
                                currentStatement = new StringBuffer();
                            }
                        }
                        else
                        {
                            currentStatement.append(sqlChars[i]);
                        }
                        break;
                    /* sybase uses 'GO' as it's statement delimiter */
                    case 'g':
                    case 'G':
                        /*
                         * Build up a string, reading backwards from the current index to
                         * the previous newline (or beginning of sequence) and from the
                         * current index up to the next newline.  If it matches the regex
                         * for the GO delimiter, then add the statement otherwise
                         * just append the current index's character to currentStatement
                         */
                        if (context.getDatabaseType().getDatabaseType().equals("sybase"))
                        {
                            // read from current index to previous line terminator 
                            // or start of sequence
                            StringBuffer previous = new StringBuffer();
                            for (int j = i - 1; j >= 0; j--) 
                            {
                                char c = sqlChars[j];
                                previous.append(c);
                                if (isLineTerminator(c))
                                {
                                    break;
                                }
                            }
                            
                            // reverse previous, since we've been walking backwards, but appending
                            previous = previous.reverse();
                            
                            // read from current index to upcoming line terminator 
                            // or end of sequence.  If it is the GO delimiter, 
                            // we skip up to line terminator
                            StringBuffer after = new StringBuffer();
                            int newIndex = 0; 
                            for (int k = i + 1; k < sqlChars.length; k++) 
                            {
                                char c = sqlChars[k];
                                after.append(c);
                                newIndex = k;
                                if (isLineTerminator(c))
                                {
                                    break;
                                }
                            }
                            
                            // check against the pattern if its a GO delimiter
                            String possibleDelimiter = previous
                                .append(sqlChars[i]).append(after).toString();
                            final String delimiterPattern = "^\\s*[Gg][Oo]\\s*$";
                            
                            if (Pattern.matches(delimiterPattern, possibleDelimiter))
                            {
                                // if it's blank, don't bother adding it since Sybase
                                // will complain about empty queries.
                                // This happens if there are two GO's with no
                                // actual SQL to run between them.
                                if (!StringUtils.isBlank(currentStatement.toString().trim()))
                                {
                                    statements.add(currentStatement.toString().trim());
                                }
                                currentStatement = new StringBuffer();
                                // skip up to next line terminator
                                i = newIndex;
                            }
                            else // not a delimiter, so just append
                            {
                                currentStatement.append(sqlChars[i]);
                            }
                        }
                        else // not a sybase db, so just append
                        {
                            currentStatement.append(sqlChars[i]);
                        }
                        break;
                    default :
                        currentStatement.append(sqlChars[i]);
                        break;
                }
            }
        }
        if (currentStatement.toString().trim().length() > 0)
        {
            statements.add(currentStatement.toString().trim());
        }
        
        return statements;
    }
    
    /**
     * Return true if the string represents a stored procedure
     * 
     * @param databaseType the type of the database
     * @param statement the statement that may be a stored procedure
     * @return true if the statement is a stored procedure for the given db type
     */
    protected boolean isStoredProcedure(String databaseType, String statement)
    {
        String currentStatement = statement.trim().toLowerCase();
        if ("oracle".equals(databaseType) 
                && (currentStatement.startsWith("begin") 
                        || currentStatement.startsWith("create or replace method") 
                        || currentStatement.startsWith("create or replace function") 
                        || currentStatement.startsWith("create or replace procedure") 
                        || currentStatement.startsWith("create or replace package")))
        {
            return true;
        }
        
        return false;
    }
    
    /**
     * return true if c is a line terminator as detailed in
     * http://java.sun.com/j2se/1.5.0/docs/api/java/util/regex/Pattern.html
     * @param c the char to test
     * @return true if it is a line terminator
     */
    protected boolean isLineTerminator(char c)
    {
        return (c == '\n') // newline
            || (c == '\r') // carriage return
            || (c == '\u0085') // next-line
            || (c == '\u2028') // line-separator
            || (c == '\u2029'); // paragraph separator
    }
    
    /**
     * Check if the current migration context is against a sybase database.
     * @param context the context to check.
     * @return true if context is in a sybase database.
     */
    protected boolean isSybase(JdbcMigrationContext context)
    {
        return context.getDatabaseType().getDatabaseType().equalsIgnoreCase("sybase");
    }

    /** {@inheritDoc} */
    public String toString()
    {
        return getName();
    }
}

File: migrate/dotnet/AutopatchNet/src/com/tacitknowledge/util/migration/ado/SqlScriptMigrationTask.cs
/* 
 * Copyright 2007 Tacit Knowledge LLC
 * 
 * Licensed under the Tacit Knowledge Open License, Version 1.0 (the "License");
 * you may not use this file except in compliance with the License. You may
 * obtain a copy of the License at http://www.tacitknowledge.com/licenses-1.0.
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#region Imports
using System;
using System.Data.Common;
using System.Collections.Generic;
using System.IO;
using System.Text;
using log4net;
using com.tacitknowledge.util.migration;
#endregion

namespace com.tacitknowledge.util.migration.ado
{
	/// <summary>
    /// Adapts a SQL (DML or DDL) database patch for use with the AutoPatch framework.
	/// </summary>
	/// <author>Scott Askew (scott@tacitknowledge.com)</author>
    /// <author>Vladislav Gangan (vgangan@tacitknowledge.com)</author>
    /// <version>$Id$</version>
    public class SqlScriptMigrationTask : MigrationTaskSupport
	{
        #region Member variables
        private static readonly ILog log = LogManager.GetLogger(typeof(SqlScriptMigrationTask));

        /// <summary>
        /// The SQL to execute
        /// </summary>
		private String sql = null;
        #endregion

        #region Constructors
		/// <summary>
        /// Creates a new <code>SqlScriptMigrationTask</code>.
		/// </summary>
		/// <param name="name">
        /// the name of the SQL script to execute; this is just an
		/// identifier and does not have to correspond to a file name 
		/// </param>
		/// <param name="level">
        /// the patch level of the migration task
		/// </param>
		/// <param name="sql">
        /// the SQL to execute
		/// </param>
		public SqlScriptMigrationTask(String name, int level, String sql)
		{
			Name = name;
			Level = level;
			this.sql = sql;
		}
		
		/// <summary>
        /// Creates a new <code>SqlScriptMigrationTask</code> containing the SQL
        /// contained in the given <code>StreamReader</code>.
		/// </summary>
		/// <param name="name">
        /// the name of the SQL script to execute; this is just an
		/// identifier and does not have to correspond to a file name 
		/// </param>
		/// <param name="level">
        /// the patch level of the migration task
		/// </param>
		/// <param name="sr">
        /// the source of the SQL to execute
		/// </param>
        /// <exception cref="IOException">if there was an IO error</exception>
        /// <exception cref="OutOfMemoryException">
        /// if there is insufficient memory to allocate a buffer for the returned string
        /// </exception>
		public SqlScriptMigrationTask(String name, int level, StreamReader sr)
		{
            Name = name;
            Level = level;

            String line = null;
            StringBuilder sb = new StringBuilder();

            while ((line = sr.ReadLine()) != null)
            {
                sb.Append(line).Append(Environment.NewLine);
            }

            sql = sb.ToString();
        }
        #endregion

        #region Public methods
        /// <seealso cref="MigrationTaskSupport.Migrate(IMigrationContext)" />
		public override void Migrate(IMigrationContext ctx)
		{
            IAdoMigrationContext context = (IAdoMigrationContext)ctx;
            String sqlStatement = null;

            try
            {
                IList<String> sqlStatements = GetSqlStatements(context);
                DbConnection dbConn = context.Connection;
                DbCommand dbCmd = null;

                // Cleaning the slate before we execute the patch.
                // This was inspired by a Sybase ASE server that did not allow
                // ALTER TABLE statements in multi-statement transactions. Instead of putting
                // an if (sybase) conditional, we decided to clean the slate for everyone.
                context.Commit();

                foreach (String tempSqlStatement in sqlStatements)
                {
                    sqlStatement = tempSqlStatement;

                    if (log.IsDebugEnabled)
                    {
                        log.Debug(Name + ": Attempting to execute: " + sqlStatement);
                    }

                    try
                    {
                        dbCmd = dbConn.CreateCommand();
                        dbCmd.CommandText = sqlStatement;
                        dbCmd.ExecuteNonQuery();
                    }
                    finally
                    {
                        if (dbCmd != null)
                        {
                            dbCmd.Dispose();
                        }

                        dbCmd = null;
                    }
                }

                context.Commit();
            }
            catch (Exception e)
            {
                String message = Name + ": Error running SQL \"" + sqlStatement + "\"";
                log.Error(message, e);

                if (e is DbException && e.InnerException != null)
                {
                    log.Error("Chained SQL Exception", e.InnerException);
                }

                throw new MigrationException(message, e);
            }
		}
		
		/// <summary>
        /// Parses the SQL (DML/DDL) to execute and returns a list of individual statements. For database
		/// types that support multiple statements in a single <code>Statement.execute</code> call,
		/// this method will return a one-element <code>List</code> containing the entire SQL
		/// file.
		/// </summary>
		/// <param name="context">
        /// the <code>IMigrationContext</code>, to figure out db type and if it can handle multiple
        /// statements at once
		/// </param>
		/// <returns>
        /// a list of SQL (DML/DDL) statements to execute
		/// </returns>
        public virtual IList<String> GetSqlStatements(IAdoMigrationContext context)
		{
			IList<String> statements = new List<String>();

			if (context.DatabaseType.MultipleStatementsSupported)
			{
				statements.Add(sql);
				return statements;
			}
			
			StringBuilder currentStatement = new StringBuilder();
			bool inQuotedString = false;
			bool inComment = false;
			char[] sqlChars = sql.ToCharArray();
			
            for (int i = 0; i < sqlChars.Length; i++)
			{
				if (sqlChars[i] == '\n')
				{
					inComment = false;
				}
				
				if (!inComment)
				{
					switch (sqlChars[i])
					{
						
						case '-': 
						case '/': 
							if (!inQuotedString && i + 1 < sqlChars.Length && sqlChars[i + 1] == sqlChars[i])
							{
								inComment = true;
							}
							else
							{
								currentStatement.Append(sqlChars[i]);
							}
							break;
						
						case '\'': 
							inQuotedString = !inQuotedString;
							currentStatement.Append(sqlChars[i]);
							break;
						
						case ';': 
							if (!inQuotedString)
							{
								// If we're in a stored procedure, just keep rolling
								if (context.DatabaseType.getDatabaseType().Equals("oracle")
                                    && (currentStatement.ToString().Trim().ToLower().StartsWith("begin")
                                        || currentStatement.ToString().Trim().ToLower().StartsWith("create or replace method")
                                        || currentStatement.ToString().Trim().ToLower().StartsWith("create or replace function")
                                        || currentStatement.ToString().Trim().ToLower().StartsWith("create or replace procedure")
                                        || currentStatement.ToString().Trim().ToLower().StartsWith("create or replace package")))
								{
									currentStatement.Append(sqlChars[i]);
								}
								else
								{
									statements.Add(currentStatement.ToString().Trim());
									currentStatement = new StringBuilder();
								}
							}
							else
							{
								currentStatement.Append(sqlChars[i]);
							}
							break;
						
						default: 
							currentStatement.Append(sqlChars[i]);
							break;
						
					}
				}
			}

			if (currentStatement.ToString().Trim().Length > 0)
			{
				statements.Add(currentStatement.ToString().Trim());
			}
			
			return statements;
		}
		
		/// <seealso cref="Object.ToString()" />
		public override String ToString()
		{
			return Name;
		}
        #endregion
	}
}
